### PR TITLE
TorchRL replay buffer

### PR DESCRIPTION
**Pull request by vmoens**
- **amend**
- **amend**
- **amend**

#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 1]
```diff
@@ -1474,16 +1467,28 @@ def train(self):
...
+                    responses = trajectory.responses
+                    logprobs = trajectory.logprobs
+                    ref_logprobs = trajectory.ref_logprobs
+                    query_response_padding_masks = trajectory.query_response_padding_masks
+                    seq_lens = trajectory.seq_lens
+                    answers = trajectory.answers
+                    policy_version = trajectory.policy_version
+                else:
+                    # FIXME: I don't think we end up here anymore
+                    # we should not be here
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:1482]: lets delete

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 2]
```diff
@@ -1736,14 +1741,11 @@ def setup(self, cfg):
...
-        self.rollout_replay_buffer = Queue(
-            actor_options={"num_cpus": 10, "num_gpus": 0},
-            maxsize=cfg.vllm.queue_maxsize,
-        )
-        self.actor_replay_buffer = Queue(
+        self.rollout_queue = Queue(
             actor_options={"num_cpus": 10, "num_gpus": 0},
             maxsize=cfg.vllm.queue_maxsize,
         )
+        self.replay_buffer = RayReplayBuffer(storage=functools.partial(LazyStackStorage, max_size=1000), batch_size=1, remote_config={"num_cpus": 10, "num_gpus": 0})
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:1748]: wanna make queue size a config param?

Comment (felipemello1) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:1748]: same for batch_size

**Review by felipemello1 — APPROVED** [review]
just minor comments

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 3]
```diff
@@ -349,8 +353,9 @@ def run(self):
                     time_total_ref_step=time_total_ref_step,
                     time_model_running=time_model_running,
                     time_waiting_buffer=time_waiting_buffer,
-                    full_queue_data_discard=full_queue_data_discard,
-                    rollout_replay_buffer_size=rollout_replay_buffer_size,
+                    # TODO: what should we do with this? We can log the total number of elements written in the buffer instead
+                    # full_queue_data_discard=full_queue_data_discard,
+                    # rollout_queue_length=rollout_queue_length,
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:358]: is there a way to know when the buffer is full and discards data?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 2 above)
Comment (vmoens) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:1748]: the batch size here can be omitted, we can either do
```
rb = ReplayBuffer(..., batch_size=x)
```
or
```
rb = ReplayBuffer(..., batch_size=None)
rb.sample(x)
```
If you mix the 2, the second precedes over the first

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 1 above)
Comment (vmoens) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py]: done

**Review by mikaylagawarecki — COMMENTED** [review]
I seemed to observe that the successes are looking quite different between this and main, is this expected behavior due to the shuffling in the ReplayBuffer?

Screenshot 2025-03-24 at 5 44 20 PM
